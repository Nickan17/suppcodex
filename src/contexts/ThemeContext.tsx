import React, { createContext, useContext, useState, useEffect, ReactNode } from 'react';
import { useColorScheme } from 'react-native';

// Theme Context
type ThemeType = 'light' | 'dark' | 'system';

interface ThemeContextType {
  theme: ThemeType;
  isDark: boolean;
  setTheme: (theme: ThemeType) => void;
  toggleTheme: () => void;
}

const ThemeContext = createContext<ThemeContextType>({
  theme: 'system',
  isDark: false,
  setTheme: () => {},
  toggleTheme: () => {},
});

export const useTheme = () => useContext(ThemeContext);

export const ThemeProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  const systemColorScheme = useColorScheme();
  const [theme, setThemeState] = useState<ThemeType>('system');

  // Determine if we're in dark mode based on theme setting and system preference
  const isDark =
    theme === 'dark' ||
    (theme === 'system' && systemColorScheme === 'dark');

  const setTheme = (newTheme: ThemeType) => {
    setThemeState(newTheme);
  };

  const toggleTheme = () => {
    setThemeState(prev => {
      if (prev === 'light') return 'dark';
      if (prev === 'dark') return 'system';
      return 'light';
    });
  };

  return (
    <ThemeContext.Provider value={{ theme, isDark, setTheme, toggleTheme }}>
      {children}
    </ThemeContext.Provider>
  );
};

// Scan Context
interface Scorecard {
  [key: string]: any;
}

interface ScanContextType {
  upc: string | null;
  scorecard: Scorecard | null;
  setResult: ({ upc, scorecard }: { upc: string | null; scorecard: Scorecard | null }) => void;
}

const ScanContext = createContext<ScanContextType | undefined>(undefined);

interface ScanProviderProps {
  children: ReactNode;
}

export const ScanProvider: React.FC<ScanProviderProps> = ({ children }) => {
  const [upc, setUpc] = useState<string | null>(null);
  const [scorecard, setScorecard] = useState<Scorecard | null>(null);

  const setResult = ({ upc, scorecard }: { upc: string | null; scorecard: Scorecard | null }) => {
    setUpc(upc);
    setScorecard(scorecard);
  };

  return (
    <ScanContext.Provider value={{ upc, scorecard, setResult }}>
      {children}
    </ScanContext.Provider>
  );
};

export const useScan = () => {
  const context = useContext(ScanContext);
  if (context === undefined) {
    throw new Error('useScan must be used within a ScanProvider');
  }
  return context;
};