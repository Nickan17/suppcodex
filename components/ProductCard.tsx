import React from 'react';
import { View, StyleSheet, Image, TouchableOpacity, Dimensions } from 'react-native';
import { useTheme } from '@/context/ThemeContext';
import Colors from '@/constants/Colors';
import Card from './ui/Card';
import Typography from './ui/Typography';
import { Bookmark } from 'lucide-react-native';
import StatusChip from './StatusChip';
import RubricStatusBar from './RubricStatusBar';
import RemediationChip from './RemediationChip';


interface ProductCardProps {
  id: string;
  name: string;
  brand: string;
  imageUrl: string;
  score: number;
  isBookmarked?: boolean;
  onPress: () => void;
  onBookmark?: () => void;
  _meta?: { status: string; remediation?: string };
  ingredients?: string;
  supplement_facts?: string;
}

export default function ProductCard({
  id,
  name,
  brand,
  imageUrl,
  score,
  isBookmarked = false,
  onPress,
  onBookmark,
  _meta,
  ingredients,
  supplement_facts,
}: ProductCardProps) {
  const { isDark } = useTheme();
  const colors = isDark ? Colors.dark : Colors.light;
  const screenWidth = Dimensions.get('window').width;
  
  // For smaller screens, we use a more compact layout
  const isSmallScreen = screenWidth < 375;

  const getScoreColor = () => {
    if (score >= 80) return colors.goodScore;
    if (score >= 50) return colors.mediumScore;
    return colors.badScore;
  };

  const getScoreBackgroundColor = () => {
    if (score >= 80) return colors.goodScoreBg;
    if (score >= 50) return colors.mediumScoreBg;
    return colors.badScoreBg;
  };

  return (
    <TouchableOpacity activeOpacity={0.7} onPress={onPress} accessibilityRole="button">
        <Card variant="elevated" style={styles.card}>
          <View style={styles.container}>
            <Image source={{ uri: imageUrl }} style={styles.image} />
            <View style={styles.content}>
              <View>
                <Typography variant="bodySmall" color={colors.textSecondary} style={styles.brand}>
                  {brand}
                </Typography>
                <Typography 
                  variant={isSmallScreen ? "bodySmall" : "body"} 
                  weight="semibold" 
                  numberOfLines={2}
                  style={styles.name}
                >
                  {name}
                </Typography>
                <View style={styles.statusRow}>
                  <StatusChip status={_meta?.status as any ?? "manual"} />
                  {_meta?.remediation && <RemediationChip remediation={_meta.remediation} />}
                </View>
                <RubricStatusBar parsed={{ title: name, ingredients, facts: supplement_facts }} showIcons={true} />
              </View>
              <View style={styles.footer}>
                <View style={[styles.scoreContainer, { backgroundColor: getScoreBackgroundColor() }]}>
                  <Typography weight="bold" color={getScoreColor()}>
                    {score}
                  </Typography>
                </View>
                {onBookmark && (
                  <TouchableOpacity style={styles.bookmarkButton} onPress={onBookmark}>
                    <Bookmark
                      size={20}
                      color={isBookmarked ? colors.primary : colors.textSecondary}
                      fill={isBookmarked ? colors.primary : 'transparent'}
                    />
                  </TouchableOpacity>
                )}
              </View>
            </View>
          </View>
        </Card>
      </TouchableOpacity>
  );
}

const styles = StyleSheet.create({
  card: {
    marginBottom: 16,
    padding: 12,
  },
  container: {
    flexDirection: 'row',
  },
  image: {
    width: 80,
    height: 80,
    borderRadius: 8,
    marginRight: 12,
  },
  content: {
    flex: 1,
    justifyContent: 'space-between',
  },
  brand: {
    marginBottom: 2,
  },
  name: {
    marginBottom: 8,
  },
  footer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  scoreContainer: {
    paddingHorizontal: 10,
    paddingVertical: 4,
    borderRadius: 16,
    justifyContent: 'center',
    alignItems: 'center',
  },
  bookmarkButton: {
    padding: 4,
  },
  statusRow: {
    flexDirection: 'row',
    gap: 8,
    marginBottom: 8,
  },
});