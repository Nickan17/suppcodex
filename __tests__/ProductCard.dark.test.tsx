import React from 'react';
import { render } from '@testing-library/react-native';
import ProductCard from '../components/ProductCard';
import { ThemeProvider } from '../context/ThemeContext';

const mockProduct = {
  id: 'test-123',
  name: 'Test Supplement',
  brand: 'Test Brand',
  imageUrl: 'https://example.com/image.jpg',
  score: 85,
  onPress: jest.fn(),
};

describe('ProductCard Dark Mode', () => {
  it('renders correctly in light theme', () => {
    const { toJSON } = render(
      <ThemeProvider>
        <ProductCard {...mockProduct} />
      </ThemeProvider>
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly in dark theme', () => {
    const { toJSON } = render(
      <ThemeProvider>
        <ProductCard {...mockProduct} />
      </ThemeProvider>
    );
    expect(toJSON()).toMatchSnapshot();
  });
}); 