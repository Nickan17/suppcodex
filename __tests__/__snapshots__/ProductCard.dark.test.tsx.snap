// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ProductCard Dark Mode renders correctly in dark theme 1`] = `
<View
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "opacity": 1,
    }
  }
>
  <View
    style={
      [
        {
          "padding": 16,
        },
        {
          "backgroundColor": "#FFFFFF",
          "borderRadius": 16,
          "elevation": 4,
          "shadowColor": "#18181B",
          "shadowOffset": {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.1,
          "shadowRadius": 8,
        },
        {
          "marginBottom": 16,
          "padding": 12,
        },
      ]
    }
  >
    <View
      style={
        {
          "flexDirection": "row",
        }
      }
    >
      <Image
        source={
          {
            "uri": "https://example.com/image.jpg",
          }
        }
        style={
          {
            "borderRadius": 8,
            "height": 80,
            "marginRight": 12,
            "width": 80,
          }
        }
      />
      <View
        style={
          {
            "flex": 1,
            "justifyContent": "space-between",
          }
        }
      >
        <View>
          <Text
            style={
              [
                {
                  "color": "#52525B",
                  "fontFamily": "Inter-Regular",
                  "fontSize": 14,
                  "lineHeight": 20,
                },
                {
                  "marginBottom": 2,
                },
              ]
            }
          >
            Test Brand
          </Text>
          <Text
            numberOfLines={2}
            style={
              [
                {
                  "color": "#18181B",
                  "fontFamily": "Inter-SemiBold",
                  "fontSize": 16,
                  "lineHeight": 24,
                },
                {
                  "marginBottom": 8,
                },
              ]
            }
          >
            Test Supplement
          </Text>
          <View
            style={
              {
                "flexDirection": "row",
                "gap": 8,
                "marginBottom": 8,
              }
            }
          >
            <View
              style={
                [
                  {
                    "borderRadius": 12,
                    "paddingHorizontal": 8,
                    "paddingVertical": 4,
                  },
                  {
                    "backgroundColor": "#F97316",
                  },
                ]
              }
            >
              <Text
                style={
                  {
                    "color": "#FFF",
                    "fontSize": 12,
                    "fontWeight": "bold",
                  }
                }
              >
                Manual
              </Text>
            </View>
          </View>
          <View
            style={
              {
                "flexDirection": "row",
                "gap": 8,
                "marginVertical": 8,
              }
            }
          >
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "borderRadius": 12,
                    "flexDirection": "row",
                    "paddingHorizontal": 6,
                    "paddingVertical": 2,
                  },
                  {
                    "backgroundColor": "#10B981",
                  },
                ]
              }
            >
              <Text
                style={
                  {
                    "fontSize": 8,
                    "marginRight": 2,
                  }
                }
              >
                ðŸŸ¢
              </Text>
              <Text
                style={
                  {
                    "color": "#FFF",
                    "fontSize": 10,
                    "fontWeight": "bold",
                  }
                }
              >
                Title
              </Text>
            </View>
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "borderRadius": 12,
                    "flexDirection": "row",
                    "paddingHorizontal": 6,
                    "paddingVertical": 2,
                  },
                  {
                    "backgroundColor": "#EF4444",
                  },
                ]
              }
            >
              <Text
                style={
                  {
                    "fontSize": 8,
                    "marginRight": 2,
                  }
                }
              >
                ðŸ”´
              </Text>
              <Text
                style={
                  {
                    "color": "#FFF",
                    "fontSize": 10,
                    "fontWeight": "bold",
                  }
                }
              >
                Ingr.
              </Text>
            </View>
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "borderRadius": 12,
                    "flexDirection": "row",
                    "paddingHorizontal": 6,
                    "paddingVertical": 2,
                  },
                  {
                    "backgroundColor": "#EF4444",
                  },
                ]
              }
            >
              <Text
                style={
                  {
                    "fontSize": 8,
                    "marginRight": 2,
                  }
                }
              >
                ðŸ”´
              </Text>
              <Text
                style={
                  {
                    "color": "#FFF",
                    "fontSize": 10,
                    "fontWeight": "bold",
                  }
                }
              >
                Facts
              </Text>
            </View>
          </View>
        </View>
        <View
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-between",
            }
          }
        >
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "borderRadius": 16,
                  "justifyContent": "center",
                  "paddingHorizontal": 10,
                  "paddingVertical": 4,
                },
                {
                  "backgroundColor": "#ECFDF5",
                },
              ]
            }
          >
            <Text
              style={
                [
                  {
                    "color": "#10B981",
                    "fontFamily": "Inter-Bold",
                    "fontSize": 16,
                    "lineHeight": 24,
                  },
                  undefined,
                ]
              }
            >
              85
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`ProductCard Dark Mode renders correctly in light theme 1`] = `
<View
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "opacity": 1,
    }
  }
>
  <View
    style={
      [
        {
          "padding": 16,
        },
        {
          "backgroundColor": "#FFFFFF",
          "borderRadius": 16,
          "elevation": 4,
          "shadowColor": "#18181B",
          "shadowOffset": {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.1,
          "shadowRadius": 8,
        },
        {
          "marginBottom": 16,
          "padding": 12,
        },
      ]
    }
  >
    <View
      style={
        {
          "flexDirection": "row",
        }
      }
    >
      <Image
        source={
          {
            "uri": "https://example.com/image.jpg",
          }
        }
        style={
          {
            "borderRadius": 8,
            "height": 80,
            "marginRight": 12,
            "width": 80,
          }
        }
      />
      <View
        style={
          {
            "flex": 1,
            "justifyContent": "space-between",
          }
        }
      >
        <View>
          <Text
            style={
              [
                {
                  "color": "#52525B",
                  "fontFamily": "Inter-Regular",
                  "fontSize": 14,
                  "lineHeight": 20,
                },
                {
                  "marginBottom": 2,
                },
              ]
            }
          >
            Test Brand
          </Text>
          <Text
            numberOfLines={2}
            style={
              [
                {
                  "color": "#18181B",
                  "fontFamily": "Inter-SemiBold",
                  "fontSize": 16,
                  "lineHeight": 24,
                },
                {
                  "marginBottom": 8,
                },
              ]
            }
          >
            Test Supplement
          </Text>
          <View
            style={
              {
                "flexDirection": "row",
                "gap": 8,
                "marginBottom": 8,
              }
            }
          >
            <View
              style={
                [
                  {
                    "borderRadius": 12,
                    "paddingHorizontal": 8,
                    "paddingVertical": 4,
                  },
                  {
                    "backgroundColor": "#F97316",
                  },
                ]
              }
            >
              <Text
                style={
                  {
                    "color": "#FFF",
                    "fontSize": 12,
                    "fontWeight": "bold",
                  }
                }
              >
                Manual
              </Text>
            </View>
          </View>
          <View
            style={
              {
                "flexDirection": "row",
                "gap": 8,
                "marginVertical": 8,
              }
            }
          >
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "borderRadius": 12,
                    "flexDirection": "row",
                    "paddingHorizontal": 6,
                    "paddingVertical": 2,
                  },
                  {
                    "backgroundColor": "#10B981",
                  },
                ]
              }
            >
              <Text
                style={
                  {
                    "fontSize": 8,
                    "marginRight": 2,
                  }
                }
              >
                ðŸŸ¢
              </Text>
              <Text
                style={
                  {
                    "color": "#FFF",
                    "fontSize": 10,
                    "fontWeight": "bold",
                  }
                }
              >
                Title
              </Text>
            </View>
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "borderRadius": 12,
                    "flexDirection": "row",
                    "paddingHorizontal": 6,
                    "paddingVertical": 2,
                  },
                  {
                    "backgroundColor": "#EF4444",
                  },
                ]
              }
            >
              <Text
                style={
                  {
                    "fontSize": 8,
                    "marginRight": 2,
                  }
                }
              >
                ðŸ”´
              </Text>
              <Text
                style={
                  {
                    "color": "#FFF",
                    "fontSize": 10,
                    "fontWeight": "bold",
                  }
                }
              >
                Ingr.
              </Text>
            </View>
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "borderRadius": 12,
                    "flexDirection": "row",
                    "paddingHorizontal": 6,
                    "paddingVertical": 2,
                  },
                  {
                    "backgroundColor": "#EF4444",
                  },
                ]
              }
            >
              <Text
                style={
                  {
                    "fontSize": 8,
                    "marginRight": 2,
                  }
                }
              >
                ðŸ”´
              </Text>
              <Text
                style={
                  {
                    "color": "#FFF",
                    "fontSize": 10,
                    "fontWeight": "bold",
                  }
                }
              >
                Facts
              </Text>
            </View>
          </View>
        </View>
        <View
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-between",
            }
          }
        >
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "borderRadius": 16,
                  "justifyContent": "center",
                  "paddingHorizontal": 10,
                  "paddingVertical": 4,
                },
                {
                  "backgroundColor": "#ECFDF5",
                },
              ]
            }
          >
            <Text
              style={
                [
                  {
                    "color": "#10B981",
                    "fontFamily": "Inter-Bold",
                    "fontSize": 16,
                    "lineHeight": 24,
                  },
                  undefined,
                ]
              }
            >
              85
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
